artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/mirror/AMQPMirrorControllerTarget.java

```diff
@@ -0,0 +1,367 @@
+import static org.apache.activemq.artemis.protocol.amqp.connect.mirror.AMQPMirrorControllerSource.QUEUE;
+import static org.apache.activemq.artemis.protocol.amqp.connect.mirror.AMQPMirrorControllerSource.ADD_ADDRESS;
+import static org.apache.activemq.artemis.protocol.amqp.connect.mirror.AMQPMirrorControllerSource.DELETE_ADDRESS;
+import static org.apache.activemq.artemis.protocol.amqp.connect.mirror.AMQPMirrorControllerSource.CREATE_QUEUE;
+import static org.apache.activemq.artemis.protocol.amqp.connect.mirror.AMQPMirrorControllerSource.DELETE_QUEUE;
+import static org.apache.activemq.artemis.protocol.amqp.connect.mirror.AMQPMirrorControllerSource.INTERNAL_ID;
+import static org.apache.activemq.artemis.protocol.amqp.connect.mirror.AMQPMirrorControllerSource.ADDRESS_SCAN_START;
+import static org.apache.activemq.artemis.protocol.amqp.connect.mirror.AMQPMirrorControllerSource.ADDRESS_SCAN_END;
+
+public class AMQPMirrorControllerTarget extends ProtonAbstractReceiver implements MirrorController {
+
+   public static final SimpleString INTERNAL_ID_EXTRA_PROPERTY = SimpleString.toSimpleString("x-opt-INTERNAL-ID");
+
+   private static final Logger logger = Logger.getLogger(AMQPMirrorControllerTarget.class);
+
+   final ActiveMQServer server;
+
+   final RoutingContextImpl routingContext = new RoutingContextImpl(null);
+
+   Map<SimpleString, Map<SimpleString, QueueConfiguration>> scanAddresses;
+
+   public AMQPMirrorControllerTarget(AMQPSessionCallback sessionSPI,
+                                     AMQPConnectionContext connection,
+                                     AMQPSessionContext protonSession,
+                                     Receiver receiver,
+                                     ActiveMQServer server) {
+      super(sessionSPI, connection, protonSession, receiver);
+      if (logger.isDebugEnabled()) {
+         logger.debug("Adding Remote Control target", new Exception("debug"));
+      }
+      this.server = server;
+   }
+
+   @Override
+   public void flow() {
+      creditRunnable.run();
+   }
+
+   @Override
+   protected void actualDelivery(AMQPMessage message, Delivery delivery, Receiver receiver, Transaction tx) {
+      Map<Symbol, Object> annotationsMap = message.getMessageAnnotationsMap(false);
+      incrementSettle();
+
+
+      if (logger.isDebugEnabled()) {
+         logger.debug("*******************************************************************************************************************************\n" +
+                      server.getIdentity() + "::Received " + message + "\n" +
+                      "*******************************************************************************************************************************");
+      }
+      try {
+         Object eventType = annotationsMap.get(EVENT_TYPE);
+         if (eventType != null) {
+            // I'm not using fancy switch with strings for JDK compatibility, just in case
```

Comment:
I thought it was beyond 8.

but the switch didn't work with symbols anyway. I'm just removing the // about JDK compatibility

Reply:
They would need to be explict strings yes.

I hadnt noticed you were using symbol values for your annotations. I'd probably change those to strings personally. Only the keys need to be symbols.

Reply:
I will keep the way it is for now.

Reply:
Oh.. you mean the value should be strings...

I thought by using Symbols I would save some space in memory during decoding?

However it doesn't really matter as the messages will be GCed anyways. I will switch them over.

Reply:
It could save some memory, thats true...so yeah in thinking further, I guess these event types are effectively operations, somewhat constants, so yes using symbols for them is reasonable.